### PR TITLE
Use VLAs when possible

### PR DESCRIPTION
**Pull request by Rhialto**
Regarding https://github.com/ggerganov/llama.cpp/pull/918#issuecomment-1505633286 , and 
https://github.com/ggerganov/llama.cpp/pull/918#issuecomment-1505599905 and #881, here is another way: use VLAs instead of alloca, when possible.

It uses `__STDC_NO_VLA__` to detect if VLAs are unavailable, in which case it falls back to alloca. In that case the program is still not a strictly conforming C program, and the `-std=c11` argument to `gcc` may evoke warnings, but on the other hand, gcc is quite likely to support VLAs in practice.

### CODE REVIEW

File: ggml.c [hunk 1]
```diff
@@ -9464,7 +9466,11 @@ void ggml_graph_compute(struct ggml_context * ctx, struct ggml_cgraph * cgraph)
         /*.has_work  =*/ false,
         /*.stop      =*/ false,
     };
+#if __STDC_NO_VLA__
     struct ggml_compute_state * workers = n_threads > 1 ? alloca(sizeof(struct ggml_compute_state)*(n_threads - 1)) : NULL;
+#else
+     struct ggml_compute_state workers[n_threads];
```
[the remark this answers is not in the document]
Comment (Rhialto) [reply ggml.c]: Oh it seems I forgot to actually push that branch... well now I have also rebased it on master.

**Review by sw — APPROVED** [review]

**Review by ggerganov — CHANGES_REQUESTED** [review]
VLAs can still involve heap allocation.

Sorry to make a big deal of this, but [ggml](https://github.com/ggerganov/ggml#features) features state "Zero memory allocations during runtime".

Let's add `GGML_NO_ALLOCA` macro to the Makefile + CMake and if set - fallback to VLAs.

**Comment by Rhialto** [comment]
Well, like I mentioned before: another option is to stop claiming the program is Standard C (as expressed by `-std=c11` and `-std=c++11`) and instead admit it isn't: `-std=gnu11` and `-std=gnu++11`. Whichever is more important to you.

**Comment by alitariq4589** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**